### DISCUSSION TITLE
GraphQL validation pipeline

### DISCUSSION THREAD

**Discussion by klippx** [MichalLytek/type-graphql#1405] [General]
GraphQL has a pipeline which is 
- parse
- validate
- execute

Currently, afaict, `type-graphql` does all it's "validation" in the execute step, which _works_ but it is not ideal. It feels unlike canonic GraphQL to validate stuff in resolvers (a shortcut taken by authors?) since resolvers should have received already parsed and validated data.

Now, if I want to write proper GraphQL scalars to parse and validate input, how do I achieve that using `type-graphql`? Is it even possible? It feels like it should be 100% possible to infer the graphql type from it and that there should be helpers to get it, and that type-graphql shouldn't even do validation in `execute` step if this pattern is used.

But all I am getting is

> `Unable to infer GraphQL type from TypeScript reflection system. You need to provide explicit type for parameter #1 of '<resolver>' of '<ClassName>' class.`

Example, if I have custom scalar such as this:

```ts
import {
  GraphQLError,
  GraphQLScalarType,
  GraphQLScalarTypeConfig,
  Kind,
  ValueNode,
} from 'graphql';

const ALPHANUMERIC = /^[a-zA-Z0-9]$/;

const validator = (rtn: string) => ALPHANUMERIC.test(rtn);

const validate = (account: unknown, ast?: ValueNode): string => {
  if (typeof account !== 'string') {
    throw new GraphQLError(
      'can only parse String',
      ast
        ? {
            nodes: ast,
          }
        : undefined
    );
  }

  if (!validator(account)) {
    throw new GraphQLError(
      'must be alphanumeric',
      ast
        ? {
            nodes: ast,
          }
        : undefined
    );
  }

  return account;
};

export const RegistrationNumberConfig: GraphQLScalarTypeConfig<string, string> =
  {
    name: 'RegistrationNumber',
    description:
      'The series of letters and figures identifying a motor vehicle, assigned on registration and displayed on a number plate.',

    serialize(value: unknown) {
      return validate(value);
    },

    parseValue(value: unknown) {
      return validate(value);
    },

    parseLiteral(ast) {
      if (ast.kind === Kind.STRING) {
        return validate(ast.value, ast);
      }

      throw new GraphQLError(
        `Registration Number can only parse String but got '${ast.kind}'`,
        {
          nodes: [ast],
        }
      );
    },
    extensions: {
      codegenScalarType: 'string',
      jsonSchema: {
        title: 'RegistrationNumber',
        type: 'string',
        pattern: ALPHANUMERIC.source,
      },
    },
  };

export const RegistrationNumber = new GraphQLScalarType(
  RegistrationNumberConfig
);
```

**Comment by MichalLytek** [comment]
https://typegraphql.com/docs/scalars.html#custom-scalars

**Comment by klippx** [comment]
Thanks @MichalLytek! Can you elaborate if the runtime code using custom scalars hooks them into the GraphQL validation phase or if all validation still happens in the execution phase? I see that the schema gets updated with expected custom scalar, but afaict in terms of validation, it is still validated in execute phase? But maybe I have some issue in my setup (federated apollo schema with a custom scalar in a subgraph).

**Reply by MichalLytek** [reply]
Graphql scalars are outside of typegraphql control.
